// @relative-ci/core/lib/cjs/env/ci-env.js
'use strict';

var index = require('../__/env-ci/index.js');
var process_env = require('../process.env.js');
var slug = require('./git/slug.js');
var github = require('./service/github.js');

// env-ci and its dependencies are not flagged as external (not bundled)
// eslint-disable-next-line import/no-extraneous-dependencies
/**
 * Load environment variables - fallback to env-ci environment variables
 */
async function getCiEnv(config, logger) {
    const { includeCommitMessage = true } = config;
    const env = process_env();
    // env-ci environment variables
    const baseCiEnv = index();
    let ciEnv = {
        isCi: baseCiEnv.isCi,
        slug: 'slug' in baseCiEnv ? baseCiEnv.slug : undefined,
        service: 'service' in baseCiEnv ? baseCiEnv.service : undefined,
        /**
         * When running during a pull request, env-ci exposes the current branch as `prBranch`
         * and `branch` as base branch
         */
        // eslint-disable-next-line no-nested-ternary
        branch: 'prBranch' in baseCiEnv && baseCiEnv.prBranch ? baseCiEnv.prBranch : ('branch' in baseCiEnv ? baseCiEnv.branch : undefined),
        commit: 'commit' in baseCiEnv ? baseCiEnv.commit : undefined,
        pr: 'pr' in baseCiEnv ? baseCiEnv.pr : undefined,
        build: 'build' in baseCiEnv ? baseCiEnv.build : undefined,
        buildUrl: 'buildUrl' in baseCiEnv ? baseCiEnv.buildUrl : undefined,
    };
    // env-ci does not provide a slug for jenkins
    // https://github.com/semantic-release/env-ci/blob/master/services/jenkins.js#LL18
    // https://www.jenkins.io/doc/book/pipeline/jenkinsfile/#using-environment-variables
    // https://plugins.jenkins.io/git/#plugin-content-environment-variables
    if (ciEnv.service === 'jenkins' && !ciEnv.slug) {
        ciEnv.slug = slug.getSlugFromGitURL(env.JENKINS_GIT_URL);
    }
    // GitHub extra data
    if (env.GITHUB_EVENT_PATH) {
        const gitHubEnv = await github.getGitHubEnv(env.GITHUB_EVENT_PATH, { includeCommitMessage }, logger);
        ciEnv = { ...ciEnv, ...gitHubEnv };
    }
    return ciEnv;
}
function isCi() {
    const baseCiEnv = index();
    return baseCiEnv.isCi;
}

exports.getCiEnv = getCiEnv;
exports.isCi = isCi;
