// @relative-ci/core/lib/cjs/ingest/ingest.js
'use strict';

var fs = require('node:fs/promises');
var path = require('node:path');
var locales_en = require('../locales/en.js');
var debug = require('../utils/debug.js');
require('../_virtual/index.js');
require('../__/signal-exit/dist/mjs/index.js');
require('../__/get-stream/source/array-buffer.js');
require('../_virtual/index2.js');
require('../__/execa/lib/promise.js');
require('../__/execa/lib/verbose.js');
var maskObjectProperty = require('../utils/mask-object-property.js');
var logger = require('../utils/logger.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var fs__default = /*#__PURE__*/_interopDefault(fs);
var path__default = /*#__PURE__*/_interopDefault(path);

async function ingest(data, params, config = {}, logger$1 = logger.logger) {
    const { key, endpoint, branch, build, buildUrl, commit, pr, slug, service, commitMessage, agentVersion, } = params;
    const payload = {
        key,
        project: slug,
        service,
        agentVersion,
        job: {
            commit,
            branch,
            prNumber: pr,
            buildNumber: build,
            buildUrl,
            commitMessage,
        },
        rawData: data,
    };
    const formattedPayload = maskObjectProperty.maskObjectProperties(payload, ['key']);
    debug.debug('Payload size', Buffer.byteLength(JSON.stringify(payload)));
    const { payloadFilepath } = config;
    /**
     * Save payload on disk for debugging
     */
    if (payloadFilepath) {
        logger$1.info('Save payload to', payloadFilepath);
        try {
            const payloadBaseDirectory = path__default.default.dirname(payloadFilepath);
            await fs__default.default.mkdir(payloadBaseDirectory, { recursive: true });
            await fs__default.default.writeFile(payloadFilepath, JSON.stringify(formattedPayload, null, 2));
        }
        catch (err) {
            // On error, catch and log
            logger$1.warn(`Error saving payload to "${payloadFilepath}"`, err instanceof Error ? err.message : undefined);
        }
    }
    logger$1.log('Send bundle stats to RelativeCI', `branch=${branch}`, `commit=${commit}`);
    let responseData = null;
    try {
        const response = await fetch(endpoint, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json; charset=utf-8',
            },
            body: JSON.stringify(payload),
        });
        responseData = await response.json();
        debug.debug('Response', responseData);
    }
    catch (error) {
        throw new Error(locales_en.INGEST_ERROR, { cause: error });
    }
    if (!responseData) {
        throw new Error(locales_en.INGEST_INVALID_DATA);
    }
    if (responseData.code) {
        throw new Error(responseData.message, { cause: responseData });
    }
    if (!responseData.res) {
        throw new Error(locales_en.INGEST_INVALID_DATA, { cause: responseData });
    }
    return responseData;
}

module.exports = ingest;
